diff --git a/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemOps.td b/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemOps.td
--- a/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemOps.td
+++ b/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemOps.td
@@ -13,7 +13,6 @@
 include "mlir/Interfaces/SideEffectInterfaces.td"
 include "mlir/Interfaces/DestinationStyleOpInterface.td"
 include "mlir/IR/CommonTypeConstraints.td"
-include "mlir/Interfaces/CopyOpInterface.td"
 include "hexagon/Dialect/HexagonMem/IR/HexagonMemDialect.td"
 include "hexagon/Dialect/HexagonMem/IR/HexagonMemTypes.td"

@@ -66,7 +65,7 @@
 // CopyOp
 //===----------------------------------------------------------------------===//

-def HexagonMem_CopyOp : HexagonMem_Op<"copy", [CopyOpInterface, SameOperandsElementType,
+def HexagonMem_CopyOp : HexagonMem_Op<"copy", [SameOperandsElementType,
     SameOperandsShape]> {

   let description = [{
diff --git a/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemDialect.h b/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemDialect.h
--- a/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemDialect.h
+++ b/qcom_hexagon_backend/include/hexagon/Dialect/HexagonMem/IR/HexagonMemDialect.h
@@ -18,7 +18,6 @@
 #include "mlir/IR/SymbolTable.h"
 #include "mlir/IR/TypeSupport.h"
 #include "mlir/IR/Types.h"
-#include "mlir/Interfaces/CopyOpInterface.h"
 #include "mlir/Interfaces/DestinationStyleOpInterface.h"

 //===----------------------------------------------------------------------===//
