diff --git a/qcom_hexagon_backend/lib/Transforms/ExpandMathOpsPass.cpp b/qcom_hexagon_backend/lib/Transforms/ExpandMathOpsPass.cpp
index f8a5242..d492b21 100644
--- a/qcom_hexagon_backend/lib/Transforms/ExpandMathOpsPass.cpp
+++ b/qcom_hexagon_backend/lib/Transforms/ExpandMathOpsPass.cpp
@@ -35,24 +35,10 @@ struct ExpandMathOpsPass : public ::impl::ExpandMathOpsBase<ExpandMathOpsPass> {
   }

   void runOnOperation() override {
-    auto moduleOp = getOperation();
-    RewritePatternSet patterns(moduleOp.getContext());
-    mlir::populateExpandTanPattern(patterns);
-    mlir::populateExpandSinhPattern(patterns);
-    mlir::populateExpandCoshPattern(patterns);
-    mlir::populateExpandTanhPattern(patterns);
-    mlir::populateExpandAsinhPattern(patterns);
-    mlir::populateExpandAcoshPattern(patterns);
-    mlir::populateExpandAtanhPattern(patterns);
-    mlir::populateExpandRsqrtPattern(patterns);
-    mlir::populateExpandCtlzPattern(patterns);
-    mlir::populateExpandFmaFPattern(patterns);
-    mlir::populateExpandCeilFPattern(patterns);
-    mlir::populateExpandExp2FPattern(patterns);
-    mlir::populateExpandPowFPattern(patterns);
-    mlir::populateExpandFPowIPattern(patterns);
-    mlir::populateExpandRoundFPattern(patterns);
-    mlir::populateExpandRoundEvenPattern(patterns);
+    // Note: This has been updated because of a deprecated API
+    // The new API is a 1:1 match with the additional math.clampf pattern
+    RewritePatternSet patterns(&getContext());
+    math::populateExpansionPatterns(patterns);
     if (failed(applyPatternsGreedily(getOperation(), std::move(patterns)))) {
       return signalPassFailure();
     }
diff --git a/qcom_hexagon_backend/lib/Transforms/VectorLoweringPass.cpp b/qcom_hexagon_backend/lib/Transforms/VectorLoweringPass.cpp
index 390e9e1..28ab902 100644
--- a/qcom_hexagon_backend/lib/Transforms/VectorLoweringPass.cpp
+++ b/qcom_hexagon_backend/lib/Transforms/VectorLoweringPass.cpp
@@ -63,7 +63,11 @@ static void doVectorLowering(ModuleOp moduleOp) {
   vector::populateScalarVectorTransferLoweringPatterns(
       patterns, /*benefit=*/1, /*allowMultipleUses=*/true);
   vector::populateVectorTransferPermutationMapLoweringPatterns(patterns);
-  vector::populateVectorMultiReductionLoweringPatterns(
+  vector::populateVectorMultiReductionReorderPatterns(
+      patterns, vectorMultiReductionLowering);
+  vector::populateVectorMultiReductionFlatteningPatterns(
+      patterns, vectorMultiReductionLowering);
+  vector::populateVectorMultiReductionUnrollingPatterns(
       patterns, vectorMultiReductionLowering);
   populateVectorTransferFullPartialPatterns(patterns, vectorTransformOptions);
   FrozenRewritePatternSet frozenPatterns = std::move(patterns);
